### DISCUSSION TITLE
Are generators like controlers updatable too ?

### DISCUSSION THREAD

**Discussion by lalmei** [fastapi-mvc/fastapi-mvc#243] [General]
The reason I ask since they depend on functions that are defined in the `Generator` class, for example, https://github.com/fastapi-mvc/fastapi-mvc/blob/02387633cc72d38fb372d023f6f675a055ce9f37/fastapi_mvc/core.py#L321 

Shouldn't this live in the controller template itself? Also, it is unclear to me if the `fastapi-mvc update` will run the generators updates too.

**Comment by rszamszur** [comment]
The `fastapi-mvc update` command will update any already generated project (`fastapi-mvc new ...`) with changes from the new template ([copier-project](https://github.com/fastapi-mvc/copier-project)) version. The correctness of this process [is being tested](https://github.com/fastapi-mvc/fastapi-mvc/blob/master/tests/integration/cli/test_update.py) the following way:
1. Generate a project from each [copier-project](https://github.com/fastapi-mvc/copier-project) release (except the latest).
2. Generate reference project from the latest release
3. For each generated project from point no. 1, run `fastapi-mvc update`
4. Compare directory checksums for updated and reference projects. If the checksums are the same, the update succeeded up to a single bit.

However, there are some corner cases where the update will not work as expected. I'm planning on improving/fixing that. But first, I'll need to discuss a few behaviors with the upstream project (copier).

Regarding functions defined in the `Generator` class. Each custom or built-in generator depends on `fastapi-mvc` package (needs `Generator` class). This provides two things. First, standardizing usage across all implementations, two providing common utilities/methods for the user. For an instance method for ensuring correct permissions to a given path, or validating fastapi-mvc project. However, now that you mentioned this, `insert_router_import` method feels more specific to the controller generator rather than something common. Also, there is a small difference between built-ins and custom generators. For the builtins, interface implementation is separated from the copier template. Example for controller generator:
* Template: https://github.com/fastapi-mvc/copier-controller
* Interface implementation: https://github.com/fastapi-mvc/fastapi-mvc/blob/master/fastapi_mvc/generators/controller.py

Maybe for builtins, this should not be separated too? But then is the question of testing and packaging the implementation for fastapi-mvc. Do you have any suggestions on this?

Lastly, any generator can be updated with the changes from the [upstream (copier-generator)](https://github.com/fastapi-mvc/copier-generator), example: 
* https://github.com/fastapi-mvc/copier-controller#updating https://github.com/fastapi-mvc/copier-controller/pull/2

But this action will not update/override your template and its configuration but rather generators common files like dotfiles, environment definition, etc. I've tested this recently, and it works well.

**Comment by lalmei** [comment]
Yeah, I think it comes down to having the base template and generators modifying the same files. There should be a way to create an intermittent template and supplement with the updated generator template., You might need to keep track of the order the generators were applied.
